### PR TITLE
Added `(get-unsat-assumptions)` to SMT solver interface

### PR DESCRIPTION
**Pull request by michael-zhang-lambda**
# Implementation
- Added `get_unsat_assumptions` to `SolverContext` and wrote function to execute query
- Added parsing function for `(get-unsat-assumptions)` solver output
- Derived `Clone` for `Lexer` in order to be able to "peek" at current, non-comment token

# Testing
Implemented five tests for `get_unsat_assumptions`:
- `UNSAT` query where all assumed facts are used (in the `UNSAT` proof)
- `UNSAT` query where only some assumed facts are used
- Trivial `UNSAT` query where `false` is asserted
- Simulated activation literal `UNSAT` query
- `UNSAT` query where facts used for `UNSAT` proof are not in `check-sat-assuming`, leading to empty `UNSAT` assertions list

# Known Issues
- Calling `get_unsat_assumptions` after a non-`UNSAT` query leads to undefined behavior. For simplicity, previous query result is not checked, so this must be enforced as a precondition manually.

### CODE REVIEW

File: patronus/src/smt/parser.rs [hunk 1]
```diff
@@ -74,6 +76,27 @@ type Result<T> = std::result::Result<T, SmtParserError>;
 
 type SymbolTable = FxHashMap<String, ExprRef>;
 
+pub fn collect_symbols(ctx: &Context, root: ExprRef, st: &mut SymbolTable) {
```
Comment (ekiwi) [inline patronus/src/smt/parser.rs]: https://github.com/cucapra/patronus/blob/6fa82c3f3b4671ef6524023e3e4bd5a57d78ebb1/patronus/src/expr/analysis.rs#L29

File: patronus/src/smt/parser.rs [hunk 2]
```diff
@@ -266,6 +289,52 @@ pub fn parse_get_value_response(ctx: &mut Context, input: &[u8]) -> Result<ExprR
     Ok(expr)
 }
 
+/// Extracts symbols from SMT solver response after `(get-unsat-assumptions)` call
```
Comment (ekiwi) [inline patronus/src/smt/parser.rs:296]: Instead of this comment with low information density, could you add a informal description of the expected pattern that you are parsing? (see the comment on the `‎parse_get_value_response` function for an example)

File: patronus/src/smt/parser.rs [hunk 3]
```diff
@@ -854,6 +923,7 @@ lazy_static! {
 // Lexer
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+#[derive(Clone)]
```
Comment (ekiwi) [inline patronus/src/smt/parser.rs]: I would prefer if you did not have to clone the lexer. Can you describe why exatly this is necessary? Maybe there is a nicer way we can solve this.

File: patronus/src/smt/solver.rs [hunk 4]
```diff
@@ -109,6 +115,14 @@ pub trait SolverContext: SolverMetaData {
     fn push(&mut self) -> Result<()>;
     fn pop(&mut self) -> Result<()>;
     fn get_value(&mut self, ctx: &mut Context, e: ExprRef) -> Result<ExprRef>;
+
+    /// # Preconditions
+    /// * Must have preceding `UNSAT` query, else behavior is undefined
+    fn get_unsat_assumptions(
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: Let's remove this comment and add an internal way of tracking whether this precondition is fulfilled.
Also, isn't the precondition, that there was an UNSAT `check-sat-assuming` call? I.e., if you did a `check-sat` call that returns in UNSAT, that would not be enough, right?

File: patronus/src/smt/solver.rs [hunk 5]
```diff
@@ -390,15 +415,40 @@ impl SolverContext for SmtLibSolverCtx {
         let expr = parse_get_value_response(ctx, response.as_bytes())?;
         Ok(expr)
     }
+
+    fn get_unsat_assumptions(
+        &mut self,
+        ctx: &mut Context,
+        props: impl IntoIterator<Item = ExprRef>,
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: I don't understand why you need `props` here. The SMT solver should know the properties and return them.

File: patronus/src/smt/solver.rs [hunk 6]
```diff
@@ -390,15 +415,40 @@ impl SolverContext for SmtLibSolverCtx {
...
+            response.as_bytes(),
+        )?)
+    }
 }
 
 pub const BITWUZLA: SmtLibSolver = SmtLibSolver {
     name: "bitwuzla",
     args: &[],
-    options: &["incremental", "produce-models"],
+    options: &["incremental", "produce-models", "produce-unsat-assumptions"],
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs:472]: We should try to see if we can make this option optional. However, this can wait until a follow up PR.

File: patronus/src/smt/solver.rs [hunk 7]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
 
+    /// Check that asserting `a == 3` and `a == 4` requires both facts to prove `UNSAT`
+    #[test]
+    fn test_bitwuzla_unsat_assumptions_basic() {
+        let mut ctx = Context::default();
+        let a = ctx.bv_symbol("a", 3);
+        let eq3 = ctx.build(|c| c.equal(a, c.bit_vec_val(3, 3)));
+        let eq4 = ctx.build(|c| c.equal(a, c.bit_vec_val(4, 3)));
+
+        let replay = Some(std::fs::File::create("replay.smt").unwrap());
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: It is better if we do not use replay files in unittest. If someone was to debug the test, they could always enable it temporarily.

File: patronus/src/smt/solver.rs [hunk 8]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
+        let res = solver.check_sat_assuming(&ctx, [eq3, eq4]).unwrap();
+        assert_eq!(res, CheckSatResponse::Unsat);
+
+        let core = solver.get_unsat_assumptions(&mut ctx, [eq3, eq4]).unwrap();
+        assert_eq!(core.len(), 2);
+        assert!(core.contains(&eq3));
+        assert!(core.contains(&eq4));
+    }
+
+    /// Check that asserting `false` initially is enough to prove `UNSAT`
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs:562]: Is this test actually stable? I.e. wouldn't it be OK for the SMT solver to return `eq3` and `eq4` instead of `smt_false`?

File: patronus/src/smt/solver.rs [hunk 9]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
+        solver.declare_const(&ctx, b).unwrap();
+
+        let res = solver.check_sat_assuming(&ctx, [eq3, eq4, b_is_1]).unwrap();
+        assert_eq!(res, CheckSatResponse::Unsat);
+
+        let core = solver
+            .get_unsat_assumptions(&mut ctx, [eq3, eq4, b_is_1])
+            .unwrap();
+        assert!(core.contains(&eq3) && core.contains(&eq4));
+        assert!(!core.contains(&b_is_1)); // the unrelated assumption is not in the core
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: It might be better to just assert that the core is exactly `2` long.

File: patronus/src/smt/solver.rs [hunk 10]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
+    }
+
+    /// Simulate activation literal `UNSAT` assumptions by asserting
+    /// `x == 2`, `x >= 5`, and `x <= 1` to yield `UNSAT` proof with only first two facts
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: This test might also be unstable since the UNSAT core is not unique, right?

File: patronus/src/smt/solver.rs [hunk 11]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
+    #[test]
+    fn test_bitwuzla_unsat_assumptions_act_lits() {
+        let mut ctx = Context::default();
+        let x = ctx.bv_symbol("x", 3);
+        let eq2 = ctx.build(|c| c.equal(x, c.bit_vec_val(2, 3)));
+        let ge3 = ctx.build(|c| c.greater_or_equal(x, c.bit_vec_val(5, 3)));
+        let le1 = ctx.build(|c| c.not(c.greater(x, c.bit_vec_val(1, 3))));
+
+        let replay = Some(std::fs::File::create("replay.smt").unwrap());
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: no replay

File: patronus/src/smt/solver.rs [hunk 12]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
+    fn test_bitwuzla_unsat_assumptions_empty() {
+        let mut ctx = Context::default();
+        let a = ctx.bv_symbol("a", 3);
+        let b = ctx.bv_symbol("b", 3);
+
+        let eq3 = ctx.build(|c| c.equal(a, c.bit_vec_val(3, 3)));
+        let eq4 = ctx.build(|c| c.equal(a, c.bit_vec_val(4, 3)));
+        let b_is_1 = ctx.build(|c| c.equal(b, c.bit_vec_val(1, 3)));
+
+        let replay = Some(std::fs::File::create("replay.smt").unwrap());
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: remove replay

File: patronus/src/smt/solver.rs [hunk 13]
```diff
@@ -463,6 +516,135 @@ mod tests {
...
+
+        let core = solver
+            .get_unsat_assumptions(&mut ctx, act_lits.clone())
+            .unwrap();
+        assert_eq!(core.len(), 2);
+        assert!(core.contains(&act_lits[0]) && core.contains(&act_lits[1]));
+    }
+
+    /// Create an `UNSAT` query with an empty `UNSAT` assumptions
+    /// Make sure that nothing crashes
```
Comment (ekiwi) [inline patronus/src/smt/solver.rs]: Not crashing is normally implied. The thing that seems interesting about this test is the fact that the unsat core is empty because the preceeding assumptions were already unsat.

**Review by ekiwi — CHANGES_REQUESTED** [review]
Lots of small comments.

The three big things are:
1. Why does `get_unsat_assumptions` need to know `props`? Can we get rid of that?
2. Why did you need to clone the `Lexer`, is there a way around this?
3. Can you add a variable inside the smt solver struct that tracks whether there was a call to `check-sat-assuming` and whether the result was unsat? Then you can check in `get_unsat_assumptions` whether it is valid to call.

File: patronus/src/smt/parser.rs (hunk 2 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/parser.rs:296]: Thanks for catching this. I've added some more descriptive documentation to this function.

File: patronus/src/smt/parser.rs (hunk 3 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/parser.rs]: Suppose we have the following SMT:

```
(set-option :print-success false)
(set-option :produce-unsat-assumptions true)
(set-logic QF_LIA)

(declare-const x Int)

; Permanent constraints
(assert (> x 10))

; Define named boolean literals to act as assumptions
(declare-const asm1 Bool)
(declare-const asm2 Bool)

; Bind the assumptions to tracking formulas
(assert (= asm1 (< x 5)))   ; Conflicting constraint
(assert (= asm2 (> x 20)))  ; Non-conflicting constraint

; Check satisfiability under these specific assumptions
(check-sat-assuming ((< x 5) (> x 20))) ; returns 'unsat'

; Query which assumptions forced the unsat result
(get-unsat-assumptions) ; returns '(asm1)'
(exit)
```

In this case, the solver returns

```
unsat                                                                                                                                                                                                                                   
((< x 5))   
```
I first consume the outer `(` token. Then, when I encounter another `(` token, I know that there is a compound SMT expression that I need to parse and I call `parse_expr_internal` to do this. However, the current implementation of `Lexer` consumed the second `(` already. Therefore, when `parse_expr_internal` starts parsing, it claims that I put a `)` without a corresponding `(`. This can be solved by allowing a "peek" feature in the `Lexer` to prevent the consumption of the first non-comment token. However, to keep things simple and avoid changing the `Lexer` implementation, I decided that it would be best to just clone `Lexer`, consume and read a token on that copy, and then consume tokens on the original `Lexer` as necessary.

Comment (michael-zhang-lambda) [reply patronus/src/smt/parser.rs]: I can probably take a look in `Lexer` and implement the "peek" feature, if you wish

File: patronus/src/smt/solver.rs (hunk 4 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: > Let's remove this comment and add an internal way of tracking whether this precondition is fulfilled. Also, isn't the precondition, that there was an UNSAT `check-sat-assuming` call? I.e., if you did a `check-sat` call that returns in UNSAT, that would not be enough, right?

Correct. This fact is somewhat highlighted in the empty `(get-unsat-assumptions)` test where we don't put the assumptions that the `UNSAT` proof used into the `(check-sat-assuming)` call. Therefore, your stronger precondition is definitely needed to prevent issues. 

I can definitely add a field to `SmtLibSolverContext` that tracks the last query result to make sure that at least part of the precondition is checked. This is definitely good because the API currently starts hanging when you try to call `(get-unsat-assumptions)` after a non-`UNSAT` query.

File: patronus/src/smt/solver.rs (hunk 5 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: Plan to resolve with internal symbol table in SMT solver infrastructure

File: patronus/src/smt/solver.rs (hunk 7 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: Removed

File: patronus/src/smt/solver.rs (hunk 9 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: In this case, I think that just asserting the core is only length-2 might cause some problems. This is because the returned core could technically be `eq_3` and `b_is_1`, which is `SAT` (since `a == 3 /\ b == 1` is valid). Therefore, this test wants to explicitly confirm that the correct facts for an `UNSAT` proof are actually returned by the solver.

File: patronus/src/smt/solver.rs (hunk 10 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: You're correct! I think this can be fixed by changing `x <= 1` to `x >= 1`.

File: patronus/src/smt/solver.rs (hunk 11 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: Removed

File: patronus/src/smt/solver.rs (hunk 12 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: Removed

File: patronus/src/smt/solver.rs (hunk 13 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs]: This is a corner case test for the parser in case that a `(get-unsat-assumptions)` call is truly empty. This occurs in this test because we didn't do a `(check-sat-assuming)` call with the relevant assumptions used in the `UNSAT` proof.

File: patronus/src/smt/solver.rs (hunk 8 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/solver.rs:562]: You're correct. I updated the test so that this ambiguity is removed.

File: patronus/src/smt/parser.rs (hunk 1 above)
Comment (michael-zhang-lambda) [reply patronus/src/smt/parser.rs]: Plan to resolve with an internal symbol table in the SMT solver interface

File: patronus/src/smt/solver.rs (hunk 9 above)
Comment (ekiwi) [reply patronus/src/smt/solver.rs]: I only meant to replace the second assertion. The other option would be to first `.sort()` the core, then you can just make it a singe `assert_eq!`

**Comment by michael-zhang-lambda** [comment]
# Changes
- Added internal symbol tables to `SmtSolver` struct to track used symbols in each context
- Added Boolean flag to `SmtSolver` struct to track whether last SAT query was `UNSAT`; allows `get_unsat_assumptions` to return error if last query was not `UNSAT` instead of resorting to undefined behavior
- Added `peek_no_comment` function to `Lexer` to prevent tokens from being consumed when checking first non-comment token; removed `#[derive(Clone)]` from `Lexer`
- Added tests for parsing `(get-unsat-assumptions)` responses (including for parsing simple literals, compound SMT-LIB expressions, and $n$-ary SMT-LIB expressions)
- Added tests to make sure that `get_unsat_assumptions` returns error instead of hanging (like before) when called after non-`UNSAT` query
- Added tests to make sure that variables in parent context (before `push`) remain in child contexts, and that symbols in child contexts are correctly removed after `pop`

**Comment by ekiwi** [comment]
@michael-zhang-lambda : Now that #20 is merged, could you rebase this branch on `main`?

File: patronus/src/smt/parser.rs [hunk 14]
```diff
@@ -1290,6 +1304,75 @@ mod tests {
         assert!(smt_expr.contains(&eq_xy) && smt_expr.contains(&ite_comp));
     }
 
+    #[test]
+    fn test_get_unsat_assumptions_parser() {
```
Comment (ekiwi) [inline patronus/src/smt/parser.rs]: are these test basically the same as the ones for `parse_epxr_list`? Can we remove them?

Comment (michael-zhang-lambda) [reply patronus/src/smt/parser.rs]: You're right. I'll remove these tests.

**Review by ekiwi — APPROVED** [review]
Thanks! looks great :shipit:
